app/services/email_normalizer.rb

```diff
@@ -0,0 +1,41 @@
+require 'resolv'
+
+# Class to help normalize email addresses for services like Gmail that let users
+# add extra things after a +
+class EmailNormalizer
+  attr_reader :email
+
+  # @param [#to_s] email
+  def initialize(email)
+    @email = Mail::Address.new(email.to_s)
+  end
+
+  # @return [String]
+  def normalized_email
+    if gmail?
+      before_plus, _after_plus = email.local.split('+', 2)
+      [before_plus.tr('.', ''), email.domain].join('@')
+    else
+      email.to_s
+    end
+  end
+
+  private
+
+  def gmail?
+    email.domain == 'gmail.com' || google_mx_record?
+  end
+
+  def google_mx_record?
+    return false if ENV['RAILS_OFFLINE']
+
+    mx_records(email.domain).any? { |domain| domain.end_with?('google.com') }
+  end
+
+  def mx_records(domain)
+    Resolv::DNS.open do |dns|
+      dns.getresources(domain, Resolv::DNS::Resource::IN::MX).
+        map { |r| r.exchange.to_s }
```

Comment:
Out of curiosity, what does the `.to_s` handle here? Can `exchange` be `nil`?

Reply:
`exchange` is an instance of `Resolv::DNS::Name`

quick lil demo:

```
Resolv::DNS.new.getresources('gsa.gov', Resolv::DNS::Resource::IN::MX)
=> 
[#<Resolv::DNS::Resource::IN::MX:0x0000000103ef51d8 @exchange=#<Resolv::DNS::Name: alt2.aspmx.l.google.com.>, @preference=5, @ttl=8973>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000103ef45d0 @exchange=#<Resolv::DNS::Name: alt4.aspmx.l.google.com.>, @preference=10, @ttl=8973>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000103ef3b80 @exchange=#<Resolv::DNS::Name: alt1.aspmx.l.google.com.>, @preference=5, @ttl=8973>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000103ef3108 @exchange=#<Resolv::DNS::Name: alt3.aspmx.l.google.com.>, @preference=10, @ttl=8973>,
 #<Resolv::DNS::Resource::IN::MX:0x0000000103ef2898 @exchange=#<Resolv::DNS::Name: aspmx.l.google.com.>, @preference=1, @ttl=8973>]
```


I ended up doing a *slightly* more realistic stub in 56a77c42a59106e08aae3d4920e4ea0db4bdf8c3

Reply:
I went for the most realistic stub in 4a8ff43315fbe92f20043acf00da35b02c3bab14

Reply:
Ah, I was looking at the Ruby docs and it didn't specify what `exchange` was, so wasn't clear to me it was a `Resolv::DNS::Name`

https://ruby-doc.org/stdlib-3.1.1/libdoc/resolv/rdoc/Resolv/DNS/Resource/MX.html